toolchain/parse/handle.h

```diff
@@ -0,0 +1,18 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_PARSE_HANDLE_H_
+#define CARBON_TOOLCHAIN_PARSE_HANDLE_H_
+
+#include "toolchain/parse/context.h"
+
+namespace Carbon::Parse {
+
+// Declare handlers for each parse state.
+#define CARBON_PARSE_STATE(Name) auto Handle##Name(Context& context) -> void;
```

Comment:
@jonmeow I think this is worth discussion to make sure we're all on the same page:

Previously, we defined these functions in the relevant `handle_*.cpp` files with no forward declaration, which I think was *probably* intentional. Adding the forward declarations means we get the substantial advantage of being able to detect missing `static`, but also means the `Handle*` functions are all in scope in all `handle_*.cpp` files, whereas we probably don't want them to be able to invoke each other.

Similarly in `check/`, where we had the header file already but didn't include it from the `handle_*.cpp` files.

I think this change is a good thing nonetheless. (It would also catch issues where the return type of a `Handle*` function is wrong.)

Reply:
Now that you remind, me I remember this being intentional. Sorry I forgot about that.

But it is definitely something to discuss -- not at all sure what the exact balance of tradeoffs is here. I was honestly surprised how many real issues `-Wmissing-prototypes` found (completely outside of any `handle.h` stuff) where we forgot headers and didn't use `static`. So mostly sending this to see if there is a good pattern we could follow here. Not at all attached to specifically having all the `Handle*` functions in one header though, very happy if we want some different pattern or way of approaching this.